presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java

```diff
@@ -242,18 +247,62 @@ public static long fromISO8601Date(ConnectorSession session, @SqlType("varchar(x
     @LiteralParameters("x")
     @SqlType(StandardTypes.TIME_WITH_TIME_ZONE)
-    public static long timeAtTimeZone(@SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long timeWithTimeZone, @SqlType("varchar(x)") Slice zoneId)
+    public static long timeAtTimeZone(ConnectorSession session, @SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long timeWithTimeZone, @SqlType("varchar(x)") Slice zoneId)
     {
-        return packDateTimeWithZone(unpackMillisUtc(timeWithTimeZone), zoneId.toStringUtf8());
+        return timeAtTimeZone(session, timeWithTimeZone, getTimeZoneKey(zoneId.toStringUtf8()));
     }
 
     @ScalarFunction(value = "at_timezone", hidden = true)
     @SqlType(StandardTypes.TIME_WITH_TIME_ZONE)
-    public static long timeAtTimeZone(@SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long timeWithTimeZone, @SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long zoneOffset)
+    public static long timeAtTimeZone(ConnectorSession session, @SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long timeWithTimeZone, @SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long zoneOffset)
     {
         checkCondition((zoneOffset % 60_000) == 0, INVALID_FUNCTION_ARGUMENT, "Invalid time zone offset interval: interval contains seconds");
         int zoneOffsetMinutes = (int) (zoneOffset / 60_000);
-        return packDateTimeWithZone(unpackMillisUtc(timeWithTimeZone), getTimeZoneKeyForOffset(zoneOffsetMinutes));
+        return timeAtTimeZone(session, timeWithTimeZone, getTimeZoneKeyForOffset(zoneOffsetMinutes));
+    }
+
+    private static long timeAtTimeZone(ConnectorSession session, long timeWithTimeZone, TimeZoneKey timeZoneKey)
+    {
+        DateTimeZone sourceTimeZone = getDateTimeZone(unpackZoneKey(timeWithTimeZone));
+        DateTimeZone targetTimeZone = getDateTimeZone(timeZoneKey);
+        long millis = unpackMillisUtc(timeWithTimeZone);
+
+        // STEP 1. Calculate source UTC millis in session start
+        millis += valueToSessionTimeZoneOffsetDiff(millis, session.getStartTime(), sourceTimeZone);
+
+        // STEP 2. Calculate target UTC millis in 1970
+        millis -= valueToSessionTimeZoneOffsetDiff(millis, session.getStartTime(), targetTimeZone);
+
+        // STEP 3. Make sure that value + offset is in 0 - 23:59:59.999
+        long localMillis = millis + targetTimeZone.getOffset(0);
+        // Loops up to 2 times in total
+        while (localMillis > TimeUnit.DAYS.toMillis(1)) {
+            millis -= TimeUnit.DAYS.toMillis(1);
+            localMillis -= TimeUnit.DAYS.toMillis(1);
+        }
+        while (localMillis < 0) {
+            millis += TimeUnit.DAYS.toMillis(1);
+            localMillis += TimeUnit.DAYS.toMillis(1);
+        }
+
+        return packDateTimeWithZone(millis, timeZoneKey);
+    }
+
+    // HACK WARNING!
+    // This method does calculate difference between timezone offset in current date (session start)
+    // and 1970-01-01 (same timezone). This is used to be able to avoid using fixed offset TZ id for
```

Comment:
You say "on 1970-01-01" but use `millisUtc` in the code. Why not `timeZone.getOffset(0)` ?